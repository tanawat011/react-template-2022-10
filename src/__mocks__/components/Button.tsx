export const Button = ({
  link,
  ...props
}: React.HTMLAttributes<never> & { [key: string]: unknown }) => {
  jest.fn().mockReturnValue(link)

  return (
    <button {...props} data-testid={'button'}>
      {props.children}
    </button>
  )
}
