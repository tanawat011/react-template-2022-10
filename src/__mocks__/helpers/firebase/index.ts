export const deleteCollection = jest.fn()
export const deleteDocument = jest.fn()
export const getDocument = jest.fn()
export const getAllDocuments = jest.fn()
export const setDocument = jest.fn()
export const subscribeCollection = jest.fn()
export const updateDocument = jest.fn()
