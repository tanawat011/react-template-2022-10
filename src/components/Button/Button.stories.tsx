import type { ComponentStory, ComponentMeta } from '@storybook/react'

import { Button } from '.'

export default {
  title: 'Components/Button',
  component: Button,
  parameters: {
    docs: {
      description: {
        component: 'Button description',
      },
      page: null,
    },
  },
  argTypes: {
    color: { control: { type: 'select' } },
    size: { control: { type: 'select' } },
    type: { control: { type: 'select' } },
    icon: { control: { type: 'none' } },
  },
} as ComponentMeta<typeof Button>

const Template: ComponentStory<typeof Button> = (props) => <Button label='Button' {...props} />

export const Customable = Template.bind({})

export const Primary = Template.bind({})
Primary.args = {
  color: 'primary',
}

export const Secondary = Template.bind({})
Secondary.args = {
  color: 'secondary',
}

export const Ternary = Template.bind({})
Ternary.args = {
  color: 'ternary',
}

export const Disabled = Template.bind({})
Disabled.args = {
  disabled: true,
}

const VariantTemplate: ComponentStory<typeof Button> = (props) => (
  <div className='flex'>
    <Primary label='Button' {...props} {...Primary.args} />
    <Secondary label='Button' {...props} {...Secondary.args} />
    <Ternary label='Button' {...props} {...Ternary.args} />
  </div>
)

export const Variants = VariantTemplate.bind({})
