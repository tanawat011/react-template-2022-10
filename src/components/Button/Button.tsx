import clsx from 'clsx'
import { ButtonHTMLAttributes } from 'react'
import tw from 'tailwind-styled-components'

export interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  label?: string
  size?: 'sm' | 'md' | 'lg'
  color?:
    | 'primary'
    | 'secondary'
    | 'ternary'
    | 'outline'
    | 'success'
    | 'danger'
    | 'warning'
    | 'info'
  icon?: React.ReactNode
  children?: React.ReactNode
  isLoading?: boolean
  onClick?: (e: React.MouseEvent<HTMLButtonElement>) => void
}

type FcButtonProps = React.FC<ButtonProps>

const TwButton: FcButtonProps = tw.button`
  flex
  ${(p: ButtonProps) => {
    switch (p.color) {
      case 'primary':
        return 'bg-space-blue'
      case 'secondary':
        return 'bg-secondary-500'
      case 'ternary':
        return 'bg-ternary-500'
      case 'outline':
        return 'bg-outline-500'
      case 'success':
        return 'bg-success-500'
      case 'danger':
        return 'bg-danger-500'
      case 'warning':
        return 'bg-warning-500'
      case 'info':
        return 'bg-info-500'
      default:
        return 'bg-aaa'
    }
  }}
`

export const Button: FcButtonProps = (props: ButtonProps) => {
  const { className, children, label = 'Button' } = props

  if (children) {
    return (
      <TwButton {...props} className={clsx(className, '')}>
        {children}
      </TwButton>
    )
  }

  return (
    <TwButton {...props}>
      <span>{label}</span>
    </TwButton>
  )
}
