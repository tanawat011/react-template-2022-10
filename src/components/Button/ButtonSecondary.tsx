import type { ButtonProps } from '.'

import React from 'react'

import { Button } from '.'

type ButtonSecondaryProps = Omit<ButtonProps, 'color'>

export const ButtonSecondary: React.FC<ButtonSecondaryProps> = (props) => (
  <Button color='secondary' {...props} />
)
