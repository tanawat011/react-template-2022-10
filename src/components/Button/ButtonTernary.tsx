import type { ButtonProps } from '.'

import React from 'react'

import { Button } from '.'

type ButtonTernaryProps = Omit<ButtonProps, 'color'>

export const ButtonTernary: React.FC<ButtonTernaryProps> = (props) => (
  <Button color='ternary' {...props} />
)
