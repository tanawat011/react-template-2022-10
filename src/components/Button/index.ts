export * from './Button'
export * from './ButtonPrimary'
export * from './ButtonSecondary'
export * from './ButtonTernary'
