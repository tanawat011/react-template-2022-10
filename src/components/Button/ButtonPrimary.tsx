import type { ButtonProps } from '../Button'

import React from 'react'

import { Button } from '../Button'

type ButtonPrimaryProps = Omit<ButtonProps, 'color'>

export const ButtonPrimary: React.FC<ButtonPrimaryProps> = (props) => (
  <Button color='primary' {...props} />
)
