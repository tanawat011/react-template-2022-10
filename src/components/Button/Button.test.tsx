import { render, screen } from '@testing-library/react'

import { renderWithProviders } from 'helpers/test'

import { Button } from './Button'

jest.mock('./Button', () => jest.requireActual('./Button'))

describe('<Button />', () => {
  test('should render', () => {
    render(<Button label='test' />)
    const button = screen.getByText(/test/i)
    expect(button).toBeInTheDocument()
  })

  test('match snapshot', () => {
    const { asFragment } = render(<Button label='test' />)
    expect(asFragment()).toMatchSnapshot()
  })

  test('should render with default setting', () => {
    render(<Button label='test' color='primary' />)
    const button = screen.getByTestId(/button/i)
    expect(button).toHaveStyleRule('background-color', 'rgb(24 28 35 / var(--tw-bg-opacity))')
  })

  test('should render primary button', () => {
    render(<Button label='test' color='primary' />)
    const button = screen.getByTestId(/button/i)
    expect(button).toHaveStyleRule('background-color', 'rgb(24 28 35 / var(--tw-bg-opacity))')
  })

  test('should render secondary button', () => {
    render(<Button label='test' color='secondary' />)
    const button = screen.getByTestId(/button/i)
    expect(button).toHaveStyleRule('background-color', 'rgb(23 98 193 / var(--tw-bg-opacity))')
  })

  test('should render ternary button', () => {
    render(<Button label='test' color='ternary' />)
    const button = screen.getByTestId(/button/i)
    expect(button).toHaveStyleRule('background-color', 'rgb(229 240 252 / var(--tw-bg-opacity))')
  })

  test('should render with child node', () => {
    render(<Button>button child</Button>)
    const button = screen.getByText(/button child/i)
    expect(button).toBeInTheDocument()
  })

  test('should render disabled button', () => {
    render(<Button label='test' disabled />)
    const button = screen.getByTestId(/button/i)
    expect(button).toHaveAttribute('disabled')
    expect(button).toHaveStyleRule('background-color', 'rgb(215 219 224 / var(--tw-bg-opacity))', {
      modifier: ':disabled',
    })
  })

  test('should render link button', () => {
    renderWithProviders(<Button link label='test' />)
    const button = screen.getByTestId(/button/i)
    expect(button).toHaveStyleRule('background-color', 'rgb(24 28 35 / var(--tw-bg-opacity))')
  })

  test('should render link button with child label', () => {
    renderWithProviders(
      <Button link label='test'>
        test
      </Button>,
    )
    const button = screen.getByTestId(/button/i)
    expect(button).toHaveStyleRule('background-color', 'rgb(24 28 35 / var(--tw-bg-opacity))')
  })
})
