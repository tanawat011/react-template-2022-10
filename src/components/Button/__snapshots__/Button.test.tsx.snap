// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> match snapshot 1`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 2.25rem;
  min-width: 156px;
  padding-left: 0.5rem;
  padding-right: 0.5rem;
  padding-top: 1rem;
  padding-bottom: 1rem;
  border-radius: 30px;
  font-family: ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,"Helvetica Neue",Arial,"Noto Sans",sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol","Noto Color Emoji";
  font-weight: 700;
  font-size: 0.75rem;
  line-height: 1.5rem;
  --tw-text-opacity: 1;
  color: rgb(255 255 255 / var(--tw-text-opacity));
  cursor: pointer;
  --tw-bg-opacity: 1;
  background-color: rgb(24 28 35 / var(--tw-bg-opacity));
}

.c0:hover {
  --tw-bg-opacity: 1;
  background-color: rgb(59 64 80 / var(--tw-bg-opacity));
}

.c0:disabled {
  --tw-bg-opacity: 1;
  background-color: rgb(215 219 224 / var(--tw-bg-opacity));
}

<button
    class="c0"
    data-testid="button"
    label="test"
  >
    <span>
      test
    </span>
  </button>
</DocumentFragment>
`;
